File: 0701-0800/0730-Count Different Palindromic Subsequences/0730-Count Different Palindromic Subsequences.cpp
class Solution {
public:
    int countPalindromicSubsequences(string S) {
        int n = S.size();
        int MOD = 1000000007;
        auto dp_ptr = new vector<vector<vector<int>>>(4, vector<vector<int>>(n, vector<int>(n)));
        auto& dp = *dp_ptr;
        for (int i = 0; i < n; ++i) {
            for (int k = 0; k < 4; ++k) {
                if (S[i] == 'a' + k) {
                    dp[k][i][i] = 1;
                }
            }
        }
        
        for (int l = 2; l <= n; ++l) {
            for (int i = 0; i <= n - l; ++i) {
                int j = i + l - 1;
                for (int k = 0; k < 4; ++k) {
                    char c = 'a' + k;
                    if (S[i] != c) {
                        dp[k][i][j] = dp[k][i + 1][j];
                    }
                    else if (S[j] != c) {
                        dp[k][i][j] = dp[k][i][j - 1];
                    }
                    else {
                        dp[k][i][j] = 2;
                        if (l != 2) {
                            for (int m = 0; m < 4; ++m) {
                                dp[k][i][j] = (dp[k][i][j] + dp[m][i + 1][j - 1]) % MOD;
                            }
                        }
                    }
                }
            }
        }
        
        int count = 0;
        for (int k = 0; k < 4; ++k) {
            count = (count + dp[k][0][n - 1]) % MOD;
        }
        
        return count;
    }
};

class Solution {
public:
    int countPalindromicSubsequences(string S) {
        int n = S.size();
        prev = vector<vector<int>>(n, vector<int>(4, -1));
        next = vector<vector<int>>(n, vector<int>(4, -1));
        dp = vector<vector<int>>(n, vector<int>(n));
        vector<int> last(4, -1);
        for (int i = 0; i < n; ++i) {
            last[S[i] - 'a'] = i;
            for (int j = 0; j < 4; ++j) {
                prev[i][j] = last[j];
            }
        }
        fill(last.begin(), last.end(), -1);
        for (int i = n - 1; i >= 0; --i) {
            last[S[i] - 'a'] = i;
            for (int j = 0; j < 4; ++j) {
                next[i][j] = last[j];
            }
        }
        
        return dfs(S, 0, n - 1);
    }
    
private:
    vector<vector<int>> prev, next, dp;
    int MOD = 1000000007;
    int dfs(string S, int start, int end) {
        if (start > end) {
            return 0;
        }
        
        if (dp[start][end]) {
            return dp[start][end];
        }
        
        int count = 0;
        for (int k = 0; k < 4; ++k) {
            int i = next[start][k];
            int j = prev[end][k];
            
            if (i == -1 || i > j) {
                continue;
            }
                        
            if (i == j) {
                ++count;
            }
            else {
                count += 2;
                count = (count + dfs(S, i + 1, j - 1)) % MOD;
            }                      
        }
        
        return dp[start][end] = count;
    }
};

File: 0701-0800/0730-Count Different Palindromic Subsequences/0730-Count Different Palindromic Subsequences.py
class Solution:
    def countPalindromicSubsequences(self, S: str) -> int:
        n = len(S)
        MOD = 1000000007
        dp = [[[0]*n for _ in range(n)] for _ in range(4)]
        for i in range(n):
            for k in range(4):
                if ord(S[i]) == ord('a') + k:
                    dp[k][i][i] = 1
        
        for l in range(2, 1 + n):
            for i in range(n - l + 1):
                j = i + l - 1
                for k in range(4):
                    c = chr(ord('a') + k)
                    if S[i] != c:
                        dp[k][i][j] = dp[k][i + 1][j]
                    elif S[j] != c:
                        dp[k][i][j] = dp[k][i][j - 1]
                    else:
                        dp[k][i][j] = 2
                        if l != 2:
                            for m in range(4):
                                dp[k][i][j] = (dp[k][i][j] + dp[m][i + 1][j - 1]) % MOD
                                
        return sum(dp[k][0][n - 1] for k in range(4)) % MOD

# With preprocess
class Solution2:
    def countPalindromicSubsequences(self, S: str) -> int:
        n = len(S)
        MOD = 1000000007
        prev = [[-1]*4 for _ in range(n)]
        next = [[-1]*4 for _ in range(n)]
        dp = [[0]*n for _ in range(n)]
        last = [-1]*4
        for i in range(n):
            last[ord(S[i]) - ord('a')] = i
            for j in range(4):
                prev[i][j] = last[j]
                
        last = [-1]*4
        for i in range(n - 1, -1, -1):
            last[ord(S[i]) - ord('a')] = i
            for j in range(4):
                next[i][j] = last[j]
                
        def dfs(start, end):
            if start > end:
                return 0
            if dp[start][end]:
                return dp[start][end]
            count = 0
            for k in range(4):
                i = next[start][k]
                j = prev[end][k]
                if i == -1 or i > j:
                    continue
                if i == j:
                    count = (count + 1) % MOD
                else:
                    count = (count + 2 + dfs(i + 1, j - 1)) % MOD
            dp[start][end] = count
            return count
        
        return dfs(0, n - 1)

